SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,473 @@
+			Utils.CheckArgument (!String.IsNullOrEmpty (endpoint), "Endpoint cannot be null or empty.");
+			this.endpoint = endpoint;
+			this.collectorUri = MakeCollectorUri(endpoint, protocol, method);
+			this.httpProtocol = protocol;
+			this.httpMethod = method;
+			this.sendLimit = sendLimit;
+			this.bufferSize = bufferSize;
+			this.byteLimitGet = byteLimitGet;
+			this.byteLimitPost = byteLimitPost;
+		}
+
+		/// <summary>
+		/// Adds an event payload to the database.
+		/// </summary>
+		/// <param name="payload">Payload.</param>
+		public void Add(TrackerPayload payload) {
+			this.eventStore.AddEvent (payload);
+		}
+
+		/// <summary>
+		/// Starts the Emitter; will be run every 10ms.
+		/// </summary>
+		public void Start() {
+			if (emitterTimer == null) {
+				Log.Debug("Emitter: Starting emitter.");
+				emitterTimer = new Timer(Emit, null, TIMER_INTERVAL, Timeout.Infinite );
+			}
+		}
+		
+		/// <summary>
+		/// Stops the Emitter; will also empty the queue into the database.
+		/// </summary>
+		public void Stop() {
+			if (emitterTimer != null) {
+				Log.Debug("Emitter: Stoppping emitter.");
+				emitterTimer.Change(Timeout.Infinite, Timeout.Infinite);
+				emitterTimer.Dispose();
+				emitterTimer = null;
+			}
+		}
+	
+		/// <summary>
+		/// Performs and orchestrates all sending of events from the database
+		/// and subsequent removal of events from the database.
+		/// </summary>
+		private void Emit(object state) {
+			sending = true;
+			List<EventRow> events = eventStore.GetDescEventRange (sendLimit);
+
+			// If some events were found start sending!
+			if (events.Count > 0) {
+				Log.Debug("Emitter: Event count: " + events.Count);
+				List<RequestResult> results = SendRequests (events);
+				events = null;
+
+				int success = 0;
+				int failure = 0;
+
+				foreach (RequestResult result in results) {
+					if (result.success) {
+						foreach (int id in result.rowIds) {
+							eventStore.DeleteEvent (id);
+						}
+						success += result.rowIds.Count;
+					} else {
+						failure += result.rowIds.Count;
+					}
+				}
+
+				Log.Debug(" + Successful: " + success);
+				Log.Debug(" + Failure: " + failure);
+
+				if (failure > 0 && success == 0) {
+					// All events failed; sleep for 10 seconds and then retry.
+					// - Sleep prevents thrashing of database in situations where nothing can send.
+					Log.Error("Emitter: All events failed to send; pausing emitter for ten seconds.");
+					sending = false;
+
+					if (emitterTimer != null) {
+						emitterTimer.Change(FAIL_INTERVAL, Timeout.Infinite);
+					}
+				} else {
+					if (emitterTimer != null) {
+						Log.Debug("Emitter: All events sent successfully; checking for more...");
+						Emit (null); // Recursively send all events.
+					}
+				}
+			} else {
+				if (emitterTimer != null) {
+					Log.Debug("Emitter: No more events to send; pausing emitter for five seconds.");
+					sending = false;
+					emitterTimer.Change(PAUSE_INTERVAL, Timeout.Infinite);
```

Comment:
Instead of speculatively checking whether any events have appeared every 5 seconds, we could use a condition variable to make the threads communicate directly. Then the emitter thread would wake up as soon as new events are available.

In Scala this would look something like:

```scala
// EventStore.scala
// called whenever events are added to the event store
sharedLock.synchronized {
  // increment events.Count here
  sharedLock.notify()
}
```

```scala
// Emitter.scala
while (true) {
  val latestEvents =  sharedLock.synchronized() {
    while (events.Count < 1) {sharedLock.wait()}
    // Get events
  }
  // Do stuff with latest events
}
```

There's a StackOverflow question on a similar topic with C# example code: http://stackoverflow.com/questions/15657637/condition-variables-c-net